lesson20.md

```diff
@@ -0,0 +1,319 @@
+* Example: `document.getElementById("NoExist");` returns `null`
+
+---
+
+### Recap - DOM
+
+* DOM is an API
+* It allows us to add/remove/change any element of a web page
+* We can make TODO lists and cool web shops
+* But how would we create a web page with "dynamic" data, for example today's weather?
+
+---
+
+<!-- .slide: id="WebAPI" -->
+
+# Web APIs
+
+---
+
+### Basic building blocks
+
+We need to first learn about a few concepts:
+
+1. JSON - How to exchange data
+1. REST API - How to talk with a server
+1. `fetch` API - How to connect to a server
+
+---
+
+### Object notation
+
+In JavaScript, we can put all keys of objects in quotes. This is purely optional:
+
+```js
+let me = {
+    "name": "John",
+    lastName: "Doe",
+    "hobbies": [ "Eat", "Sleep" ]
+};
+
+console.log(me.name); // "John"
+```
+
+---
+
+### JSON
+
+* JSON stands for **J**ava**S**cript **O**bject **N**otation
+* It is a string representation of a JavaScript object
+* All *keys* in our key value pairs must be in quotes
+* Must contain only these types:
+   * `string`, `number`, `boolean`, `array`
+   * `null`
+   * another JSON `object`
+* These types are **forbidden**:
+   * `function`, `undefined`
+
+---
+
+### Why JSON?
+
+* When we exchange data between a browser and a server, we can only exchange data, for example a string (not objects, arrays, numbers, booleans...)
+* JSON is a string representation of JavaScript objects
+* JSON can be easily transferred (and stored)
+* The notation is very close to JavaScript objects, easy to handle from JavaScript. No complex transformation required.
+* More info: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/JSON
+
+---
+
+### JSON example
+
+```json
+{
+    "name": "Bob",
+    "instructor": true,
+    "hobbies": [ "eat", "sleep" ],
+    "age": 86,
+    "workAddress": {
+        "street": "Invalidenstraße",
+        "number": 116,
+        "city": "Berlin"
+    }
+}
+```
+
+---
+
+### object <-> JSON
+
+JavaScript offers us an API to convert an object to JSON and vice versa:
+
+```js
+let obj = {
+    name: "Bob",
+    instructor: true
+};
+
+// turn our object to JSON:
+let json = JSON.stringify(obj);
+
+// turn our JSON back to an object:
+let jsonAsObj = JSON.parse(json);
+```
+
+---
+
+### Quiz
+
+  What does the following output?
+
+```js
+typeof JSON.stringify({ name: "Bob" });
+typeof JSON.parse('{"city":"Berlin"}');
+```
+
+---
+
+### Quiz - answer
+
+```js
+typeof JSON.stringify({ name: "Bob" }); // "string"
+typeof JSON.parse('{"city":"Berlin"}'); // "object"
+```
+
+---
+
+### Web API
+
+* API stands for **A**pplication **P**rogramming **I**nterface
+* A Web API is the *contract* between your application and a server
+
+![API](images/API.png) <!-- .element height="400px" width="400px" -->
+
+---
+
+### REST API
+
+* REST stands for **Re**presentational **S**tate **T**ransfer
+* REST APIs work via http(s), the same protocol we use to connect to any website
```

Comment:
Maybe explain in simple terms what a protocol is, by using a comparison. API allows executing a code someone else has written, while a protocol allows a "fluent conversation" between two different machines/software/hardware/applications